File: smart-hub-direct-kt/src/main/kotlin/io/github/tassiLuca/hub/application/LightingManager.kt
package io.github.tassiLuca.hub.application

import io.github.tassiLuca.hub.core.LightingSystem
import io.github.tassiLuca.hub.core.LuminosityEntry
import io.github.tassiLuca.hub.core.ports.DashboardService
import kotlinx.coroutines.flow.Flow
import kotlin.coroutines.CoroutineContext
import kotlin.time.Duration.Companion.seconds

/** Manages the thermostat and the temperature sensors. */
class LightingManager(dashboardService: DashboardService, coroutineContext: CoroutineContext) {

    /** The thermostat. */
    val lightingSystem = LightingSystem(SAMPLING_WINDOW, coroutineContext, dashboardService)

    /** Runs the thermostat and the temperature sensors. */
    suspend fun run(sensorSource: Flow<LuminosityEntry>) {
        println("LightingManager: run")
        lightingSystem.run()
        sensorSource.collect {
            println("LightingManager: $it")
            lightingSystem.react(it)
        }
    }

    companion object {
        private val SAMPLING_WINDOW = 5.seconds
    }
}

File: smart-hub-direct/src/main/scala/io/github/tassiLuca/hub/application/LightingManager.scala
package io.github.tassiLuca.hub.application

import gears.async.{Async, AsyncOperations, ReadableChannel}
import io.github.tassiLuca.hub.core.ports.{DashboardServiceComponent, LampsComponent}
import io.github.tassiLuca.hub.core.{LightingSystemComponent, LuminosityEntry}
import io.github.tassiLuca.rears.{Controller, bufferWithin}

import scala.concurrent.duration.DurationInt
import scala.language.postfixOps

/** The hub managing the lighting system. */
trait LightingManager extends LightingSystemComponent with LampsComponent with DashboardServiceComponent:
  override val lightingSystem: LightingSystem = LightingSystem()
  private val samplingWindow = 10 seconds

  /** Runs the manager, spawning a new controller consuming the given [[source]] of events. */
  def run(source: ReadableChannel[LuminosityEntry])(using Async, AsyncOperations): Unit =
    lightingSystem.asRunnable.run
    Controller.oneToOne(
      publisherChannel = source,
      consumer = lightingSystem,
      transformation = r => r.bufferWithin(samplingWindow),
    ).run
